File: app/src/main/java/com/kstechsolutions/mvvm/modules/posts/adapter/PostListAdapter.java
package com.kstechsolutions.mvvm.modules.posts.adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.kstechsolutions.mvvm.R;
import com.kstechsolutions.mvvm.modules.posts.data.IPostView;
import com.kstechsolutions.mvvm.modules.posts.model.PostAndPhoto;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * Created by muhammadumairshafique on 16/04/2018.
 */


public class PostListAdapter extends RecyclerView.Adapter<PostListAdapter.PostsViewHolder> {
    private PostAndPhoto mData;
    private Context mContext;
    private IPostView mListener;

    public PostListAdapter(Context context) {
        this.mContext = context;
    }

    @Override
    public PostListAdapter.PostsViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new PostListAdapter.PostsViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.item_post, parent, false));
    }

    @Override
    public void onBindViewHolder(PostListAdapter.PostsViewHolder holder, int position) {
        holder.tvTitle.setText(mData.mPosts.get(position).getTitle());
        Glide.with(mContext).load(mData.mPhotos.get(position).getUrl()).into(holder.imageView);
    }


    @Override
    public int getItemCount() {
        return mData != null ? mData.mPosts.size() : 0;
    }

    public void setListener(IPostView listener) {
        mListener = listener;
    }

    public class PostsViewHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.tv_title)
        TextView tvTitle;
        @BindView(R.id.imageview)
        ImageView imageView;

        private PostsViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }

        @OnClick(R.id.cl_item)
        void onBodyClick() {
            mListener.onListItemClicked(mData.mPosts.get(getAdapterPosition()).getId(),
                    mData.mPosts.get(getAdapterPosition()).getTitle(),
                    mData.mPhotos.get(getAdapterPosition()).getUrl());
        }
    }


    public void setData(PostAndPhoto data) {
        mData = data;
        notifyDataSetChanged();
    }
}

File: app/src/main/java/com/kstechsolutions/mvvm/modules/posts/PostListAdapter.kt
package com.kstechsolutions.mvvm.modules.posts

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.kstechsolutions.mvvm.data.databinding.BindableAdapter
import com.kstechsolutions.mvvm.databinding.LayPostItemBinding
import javax.inject.Inject

/**
 * Created by muhammadumairshafique on 29/09/2019.
 */

class PostListAdapter @Inject constructor() : ListAdapter<PostAndPhoto, PostListAdapter.ViewHolder>(DIFF_CALLBACK), BindableAdapter<List<PostAndPhoto>> {
    private lateinit var listener: OnItemClickListener

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val binding = LayPostItemBinding.inflate(inflater, parent, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(getItem(position), listener)
    }

    override fun setData(data: List<PostAndPhoto>) {
        submitList(data)
    }

    fun setListener(listener: OnItemClickListener) {
        this.listener = listener
    }

    inner class ViewHolder(private val binding: LayPostItemBinding) : RecyclerView.ViewHolder(binding.root) {
        fun bind(item: PostAndPhoto, listener: OnItemClickListener) {
            binding.item = item
            binding.clItem.setOnClickListener { listener.onItemClicked(item) }
        }
    }

    interface OnItemClickListener {
        fun onItemClicked(item: PostAndPhoto)
    }

    companion object {
        private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<PostAndPhoto>() {
            override fun areItemsTheSame(oldItem: PostAndPhoto, newItem: PostAndPhoto) = oldItem.post == newItem.post

            override fun areContentsTheSame(oldItem: PostAndPhoto, newItem: PostAndPhoto) = oldItem == newItem
        }
    }
}